# Tarea 1 de Introducción a la Optimización

David Antonio Garzón Avendaño 

1- ¿Qué condiciones deben cumplir la Matriz $A$ y el vector $b$ para que se pueda utilizar el Metodo de Minimos Cuadrados ?
 
2- Usando el Metodo de Minimos Cuadrados para modelos no lineales ajustelo para el precio del Bitcoin y haga una predicción para determinar el precio de la semana posterior, comparelo con los valores actuales. 

3- Use el metodo de Minimos Cuadrados para determinar la relación entre el Bitcoin y los precios del petroleo, la plata y el oro durante un año.  



## Punto 1

Si en la matriz $A$ todas sus columnas son linealmnete independientes entonces $A^T A$ tiene inversa. 

Lema: 
$$ rango(A) = rango(A^TA)$$ 

Demostración (Contradicción):

Supongamos que todas las columnas de $A$ son linealmente independientes y que $A^T A$ no tiene inversa. Sabemos que $A$ es una matriz $mxn$ donde $m \geq n$ para que el problema tenga sentido, entonces el $rango(A) = n$, pues $n\leq m$ y todas las columnas son linealmnete independientes. Por otro lado tenemos que $A^T A$ es una matriz $nxn$ y usando el lema anterior tenemos que $rango(A^TA) = n$ por lo que podemos concluir que $A^T A$ es de rango completo =><= pues $A^T A$ no tiene inversa. 


Por otro lado, un sistema $Cx = b$ tiene solución si y solo si $b$ es una combinación lineal de las columnas de A, por lo que en este caso seria necesario que $b$ sea combinación lineal de las columnas de $A^T A$

## Punto 2

Para este punto usamos los datos del precio del Bitcoin desde el 17 de marzo del 2021 hasta el 17 de marzo del 2022. Usaremos los precios que se registraron al momento del cierre del mercado. Note que en este caso tambien es necesario invertir el orden de los precios, ya que en el dataframe el primer dato no corresponde al precio del 17 de marzo del 2021 sino al precio del 17 de marzo del 2022. 

In [1]:
using CSV
using DataFrames
using PlotlyJS


#Datos del valor del Bitcoin desde el 17 de marzo del 2021 hasta el 17 de marzo del 2022
dataBitcoin = CSV.read("BTC.csv",DataFrame,delim=',')
dataBitcoin.Price .= replace.(dataBitcoin.Price, "," => "");

#Invertimos el orden de los valores para que la regrecion tenga sentido
dataBitcoin.PriceFloat = reverse(parse.(Float64, dataBitcoin.Price)); 
date = Array((1:366));
dataBitcoin

WebIO._IJuliaInit()

366×8 DataFrame
 Row │ Date          Price     Open      High      Low       Vol.     Change % ⋯
     │ String15      String15  String15  String15  String15  String7  String7  ⋯
─────┼──────────────────────────────────────────────────────────────────────────
   1 │ Mar 17, 2022  40910.0   41,132.0  41,397.9  40,551.4  1.52B    -0.50%   ⋯
   2 │ Mar 16, 2022  41115.5   39,309.4  41,700.0  38,951.0  3.55B    4.67%
   3 │ Mar 15, 2022  39280.3   39,678.3  39,850.9  38,236.9  1.80B    -0.99%
   4 │ Mar 14, 2022  39671.4   37,778.5  39,913.8  37,617.6  1.82B    4.97%
   5 │ Mar 13, 2022  37793.6   38,798.5  39,270.0  37,607.3  1.27B    -2.63%   ⋯
   6 │ Mar 12, 2022  38815.4   38,691.7  39,358.1  38,665.5  939.49M  0.22%
   7 │ Mar 11, 2022  38729.6   39,419.7  40,175.9  38,234.4  2.31B    -1.73%
   8 │ Mar 10, 2022  39411.8   41,944.4  42,022.5  38,604.5  2.85B    -6.00%
   9 │ Mar 09, 2022  41926.1   38,703.9  42,528.1  38,663.3  2.80B    8.26%    ⋯
  10 │ Mar 08, 2022  38727.3   38,021.3  39,348.5  37,889.1  2.15B    1.85%
  11 │ Mar 07, 2022  38025.2   38,446.0  39,531.3  37,184.7  2.45B    -1.03%
  ⋮  │      ⋮           ⋮         ⋮         ⋮         ⋮         ⋮        ⋮     ⋱
 357 │ Mar 26, 2021  55022.1   51,310.9  55,022.1  51,310.9  3.40B    7.23%
 358 │ Mar 25, 2021  51310.3   52,429.0  52,929.5  51,092.6  4.54B    -1.96%   ⋯
 359 │ Mar 24, 2021  52338.5   54,270.4  56,993.9  52,338.5  4.58B    -3.63%
 360 │ Mar 23, 2021  54310.0   54,088.9  55,639.5  53,282.7  3.26B    0.23%
 361 │ Mar 22, 2021  54185.2   57,440.9  57,815.2  54,185.2  3.53B    -5.58%
 362 │ Mar 21, 2021  57385.0   58,040.4  58,364.1  56,160.0  2.77B    -1.21%   ⋯
 363 │ Mar 20, 2021  58090.0   58,007.9  59,553.7  58,007.9  2.62B    -0.01%
 364 │ Mar 19, 2021  58097.8   57,573.2  58,958.1  57,500.0  3.05B    0.78%
 365 │ Mar 18, 2021  57645.4   59,108.4  59,340.9  57,645.4  3.90B    -2.15%
 366 │ Mar 17, 2021  58914.2   56,911.6  58,914.2  54,525.1  3.96B    3.59%    ⋯
                                                   1 column and 345 rows omitted

In [2]:
dataBitcoin.PriceFloat

366-element Vector{Float64}:
 58914.2
 57645.4
 58097.8
 58090.0
 57385.0
 54185.2
 54310.0
 52338.5
 51310.3
 55022.1
 55855.8
 55756.7
 57610.1
     ⋮
 38420.8
 38025.2
 38727.3
 41926.1
 39411.8
 38729.6
 38815.4
 37793.6
 39671.4
 39280.3
 41115.5
 40910.0

In [3]:
plot(dataBitcoin.PriceFloat)

data: [
  "scatter with fields type, x, and y"
]

layout: "layout with fields margin and template"

Usaremos el metodo de minimos cuadrados para modelos no lineales. En este caso vamos a usar funciones polinomicas de grado 2 hasta 5 e iremos viendo cómo se ajustan a la gráfica. Usaremos 366 datos, por lo que nuestro polinomio será evaluado en los días 1 al 366 en la gráfica. 

In [4]:
N = size(dataBitcoin.PriceFloat,1) #Guardamos la cantidad de datos que tenemos.

days = Array(1:N) #Creamos un array con los dias con los que vamos a estimar el polinomio.

Polyvalues=[] #Guardamoes en este array los coeficientes de los polinomios
 
b = dataBitcoin.PriceFloat; #Definimos el vector b con los datos del precio del bitcoin.

Polydata =  [ones(N,1) ones(N,1)] #creamos un array para guardar los datos de nustro polinomio

#Bucle para usar el metodo de minimos cuadrados con polinomios de grado 2 hasta 5
for degree in 2:5

    #Creamos la matriz A con los dalores de days 
    A = [ones(N,1) days]
    
    #Si el grado de nuestro polinomio es mayor que 2, ejecutamos el bucle 
    if degree >= 2
        for i = 2:degree
          A = [A days.^i]  
        end   
    end 
    
    #Resolvemos el sistema y guardamos los coeficientes del polinomio para usarlos posteriormnete 
    xhat = A\b
    append!(Polyvalues,[xhat])
         
    
    poly = xhat[1]*ones(N,1)
       
    #Evaluamos cada polinomio desde el día 1 hasta el 366
    for i = 2:degree + 1
        poly = poly + xhat[i]*A[:,i]  
    end 
    
    #Guardamos en el Array Polydata la información anterior
    Polydata = [Polydata poly]  
    
end 


#Creamos el plot y renomabramos las funciones
p = PlotlyJS.plot(days, [b Polydata[:,3:6]]);
restyle!(p, name=["Bitcoin", "2 Grade Poly", "3 Grade Poly", "4 Grade Poly", "5 Grade Poly"]);


p

data: [
  "scatter with fields name, type, x, and y",
  "scatter with fields name, type, x, and y",
  "scatter with fields name, type, x, and y",
  "scatter with fields name, type, x, and y",
  "scatter with fields name, type, x, and y"
]

layout: "layout with fields margin and template"

En la gráfica anterior podemos ver que el polinomio de grado 5 parece ajustarse mejor a los datos, razón por la cual vamos a usar este modelo.


Ahora comparamos los precios del Bitcoin de esta semana, para ello cargaremos los datos de los últimos 7 días 
(estos datos son posteriores a los que se utilizaron en el modelo y pertenecen a otro CSV). En este caso tambien es necesario invertir el orden del estos datos debido a lo que se explicó al inicio. 

In [5]:
#Datos del valor del Bitcoin desde el 18 de marzo del 2022 hasta el 24 de marzo del 2022
dataBitcoinActual = CSV.read("BTC_actual.csv",DataFrame,delim=',')
dataBitcoinActual.Price .= replace.(dataBitcoinActual.Price, "," => "");
dataBitcoinActual.PriceFloat = reverse(parse.(Float64, dataBitcoinActual.Price));
dataBitcoinActual.PriceFloat

7-element Vector{Float64}:
 41774.0
 42214.0
 41249.0
 41031.0
 42370.0
 42882.0
 43940.0

Ahora usamos los coeficientes del polinomoio de grado 5 que se encuentran en la matriz Polydata (que almacena los coeficientes del modelo anterior de los polinomios de grado 2 hasta 5) y evaluamos en los días 367 al 373.

In [6]:
#Creamos un vector con los números del 367 al 373, estos serán los días en los cuales evaluaremos el polinomio. 
days_p = 367:373

#Usamos los coefientes del polinomio de grado 5, que están guardados en la posición 4 del array
Grade_five_Poly_data = Polyvalues[4][6]*days_p.^5 + Polyvalues[4][5]*days_p.^4 + Polyvalues[4][4]*days_p.^3 + 
Polyvalues[4][3]*days_p.^2 + Polyvalues[4][2]*days_p + Polyvalues[4][1]*ones(7,1)


7×1 Matrix{Float64}:
 48860.27199142676
 49866.76529058048
 50921.22325051728
 52024.73352032618
 53178.39674226364
 54383.32661859708
 55640.64997842258

Ahora le restamos el vector con los valores reales al vector con los valores predichos para los dias 367 al 373.

In [7]:
error_pred = dataBitcoinActual.PriceFloat - Grade_five_Poly_data

7×1 Matrix{Float64}:
  -7086.271991426758
  -7652.765290580479
  -9672.22325051728
 -10993.733520326183
 -10808.396742263641
 -11501.326618597079
 -11700.649978422582

En este caso la máxima diferencia entre el valor real y el predicho es de -11700.6499 dolares.

## Punto 3

Cargamos los precios del oro, la plata y el petroleo, en este caso como los precios de las commodities no se registran los fines de semana y los del Bitcoin si, por lo que fue necesario elimanar los datos de los días en los cuales se registró el precio del Bitcoin pero no se registró el precio de estas commodities.

In [8]:
#Datos del valor del Bitcoin desde el 17 de marzo del 2021 hasta el 17 de marzo del 2022
dataBitcoin2 = CSV.read("BTC.csv",DataFrame,delim=',')
dataBitcoin2.Price .= replace.(dataBitcoin2.Price, "," => "");
dataBitcoin2.PriceFloat = parse.(Float64, dataBitcoin2.Price)


#Cargamos los datos del Oro desde el 17 de marzo del 2021 hasta el 17 de marzo del 2022
dataGold = CSV.read("Gold.csv",DataFrame,delim=',')
dataGold.Price .= replace.(dataGold.Price, "," => "");
dataGold.PriceFloat = parse.(Float64, dataGold.Price);

#Cargamos los datos de la plata desde el 17 de marzo del 2021 hasta el 17 de marzo del 2022
dataSilver = CSV.read("Silver.csv",DataFrame,delim=',')
rename!(dataSilver, :Price => :PriceFloat) #estos datos ya son flotantes, por lo que renombramos la columna 

#Cargamos los datos del Petroleo desde el 17 de marzo del 2021 hasta el 17 de marzo del 2022
dataOil = CSV.read("Brent_Oil.csv",DataFrame,delim=',');
rename!(dataOil, :Price => :PriceFloat) #estos datos ya son flotantes, por lo que renombramos la columna 

inter = intersect(dataBitcoin2.Date, dataOil.Date, dataSilver.Date, dataGold.Date);

#Los precios de las comodities no se registran los fines de semana, pero los del bitcoin si.
#Se procede a filtar por las fechas que coincden en los 4 data frames

dataOil = dataOil[in(inter).(dataOil.Date), :];
dataSilver = dataSilver[in(inter).(dataSilver.Date), :];
dataGold = dataGold[in(inter).(dataGold.Date), :];
dataBitcoin2 = dataBitcoin2[in(inter).(dataBitcoin2.Date), :];

dataBitcoin2

261×8 DataFrame
 Row │ Date          Price     Open      High      Low       Vol.     Change % ⋯
     │ String15      String15  String15  String15  String15  String7  String7  ⋯
─────┼──────────────────────────────────────────────────────────────────────────
   1 │ Mar 17, 2022  40910.0   41,132.0  41,397.9  40,551.4  1.52B    -0.50%   ⋯
   2 │ Mar 16, 2022  41115.5   39,309.4  41,700.0  38,951.0  3.55B    4.67%
   3 │ Mar 15, 2022  39280.3   39,678.3  39,850.9  38,236.9  1.80B    -0.99%
   4 │ Mar 14, 2022  39671.4   37,778.5  39,913.8  37,617.6  1.82B    4.97%
   5 │ Mar 11, 2022  38729.6   39,419.7  40,175.9  38,234.4  2.31B    -1.73%   ⋯
   6 │ Mar 10, 2022  39411.8   41,944.4  42,022.5  38,604.5  2.85B    -6.00%
   7 │ Mar 09, 2022  41926.1   38,703.9  42,528.1  38,663.3  2.80B    8.26%
   8 │ Mar 08, 2022  38727.3   38,021.3  39,348.5  37,889.1  2.15B    1.85%
   9 │ Mar 07, 2022  38025.2   38,446.0  39,531.3  37,184.7  2.45B    -1.03%   ⋯
  10 │ Mar 04, 2022  39140.5   42,442.9  42,520.4  38,600.0  2.53B    -7.81%
  11 │ Mar 03, 2022  42458.5   43,903.9  44,078.7  41,848.3  2.19B    -3.29%
  ⋮  │      ⋮           ⋮         ⋮         ⋮         ⋮         ⋮        ⋮     ⋱
 252 │ Mar 30, 2021  58757.3   57,658.4  59,011.6  57,151.0  3.22B    1.99%
 253 │ Mar 29, 2021  57610.1   55,794.4  58,148.8  55,302.2  3.88B    3.32%    ⋯
 254 │ Mar 26, 2021  55022.1   51,310.9  55,022.1  51,310.9  3.40B    7.23%
 255 │ Mar 25, 2021  51310.3   52,429.0  52,929.5  51,092.6  4.54B    -1.96%
 256 │ Mar 24, 2021  52338.5   54,270.4  56,993.9  52,338.5  4.58B    -3.63%
 257 │ Mar 23, 2021  54310.0   54,088.9  55,639.5  53,282.7  3.26B    0.23%    ⋯
 258 │ Mar 22, 2021  54185.2   57,440.9  57,815.2  54,185.2  3.53B    -5.58%
 259 │ Mar 19, 2021  58097.8   57,573.2  58,958.1  57,500.0  3.05B    0.78%
 260 │ Mar 18, 2021  57645.4   59,108.4  59,340.9  57,645.4  3.90B    -2.15%
 261 │ Mar 17, 2021  58914.2   56,911.6  58,914.2  54,525.1  3.96B    3.59%    ⋯
                                                   1 column and 240 rows omitted

Ahora que los data frames están listos, procedemos a utilizar el Metodo de minimos cuadrados para relacionar el precio del Bitcoin con el precio del Oro, la Plata y el petroleo (de referencia brent). 

In [9]:
#Guardamos la cantidad de datos que tenemos.
N = size(dataBitcoin2.PriceFloat,1) 

#Definimos el vector b con los datos del precio del bitcoin.
b = dataBitcoin2.PriceFloat;

#Construimos la matriz de diseño
A = [ones(N,1) dataGold.PriceFloat dataOil.PriceFloat dataSilver.PriceFloat];

xhat = A\b 

4-element Vector{Float64}:
 167200.37937460202
    -65.99633764319621
     35.41391276603188
   -127.9527470390973

Ahora probamos el modelo con algunos datos (los del 17 de marzo del 2022), para este fecha el precio del Bitcoin fue: 40910.0

In [10]:
#Creamos una función con los pesos anteriores
relation(Gold, Oil, Silver) = Gold*xhat[2] + Oil*xhat[3] + Silver*xhat[4] + xhat[1]

res = relation(1943.20 ,106.64, 25.616)

39455.198155559265

Ahora usamos nuestro modelo con los datos de las commodities de la semana actual (del 18 de marzo del 2022 al 24 de marzo del 2022) y comparamos con el precio del Bitcoin en el mismo lapso de tiempo. 

In [11]:
#Cargamos los datos del Oro de esta semana
dataGoldActual = CSV.read("Gold_actual.csv",DataFrame,delim=',')
dataGoldActual.Price .= replace.(dataGoldActual.Price, "," => "");
dataGoldActual.PriceFloat = parse.(Float64, dataGoldActual.Price);

#Cargamos los datos de la plata de esta semana
dataSilverActual = CSV.read("Silver_actual.csv",DataFrame,delim=',')
rename!(dataSilverActual, :Price => :PriceFloat);

#Cargamos los datos del Petroleo de esta semana
dataOilActual = CSV.read("Brent_oil_actual.csv",DataFrame,delim=',');
rename!(dataOilActual, :Price => :PriceFloat);

#Cargamos los datos del Bitcoin de esta semana
dataBitcoinActual2 = CSV.read("BTC_actual.csv",DataFrame,delim=',');
dataBitcoinActual2.Price .= replace.(dataBitcoinActual2.Price, "," => "");
dataBitcoinActual2.PriceFloat = parse.(Float64, dataBitcoinActual2.Price);



inter2 = intersect(dataBitcoinActual2.Date, dataOilActual.Date, dataSilverActual.Date, dataGoldActual.Date);
dataOilActual = dataOilActual[in(inter2).(dataOilActual.Date), :];
dataGoldActual = dataGoldActual[in(inter2).(dataGoldActual.Date), :];
dataSilverActual = dataSilverActual[in(inter2).(dataSilverActual.Date), :];
dataBitcoinActual2 = dataBitcoinActual2[in(inter2).(dataBitcoinActual2.Date), :];



Utilzamos el modelo anterior con los valores actuales del Oro, la plata y el petroleo (del 18 de marzo de 2022 al 24 de marzo del 2022) para ver qué tan cerca está nuestro modelo del valor real del bitcoin

In [12]:
relatio_next_five_days = dataGoldActual.PriceFloat*xhat[2] + dataOilActual.PriceFloat*xhat[3] + 
dataSilverActual.PriceFloat*xhat[4] + xhat[1]*ones(5,1)

5×1 Matrix{Float64}:
 38731.18764738162
 40429.004505619654
 41291.480027160185
 40716.13460026287
 40485.91819945155

Mostramos el valor real del Bitcoin

In [13]:
dataBitcoinActual2.PriceFloat

5-element Vector{Float64}:
 43940.0
 42882.0
 42370.0
 41031.0
 41774.0

Calculamos la diferencia entre el valor estimado y el valor real del Bitcoin

In [14]:
error_est = dataBitcoinActual2.PriceFloat - relatio_next_five_days

5×1 Matrix{Float64}:
 5208.812352618377
 2452.995494380346
 1078.5199728398147
  314.86539973712934
 1288.081800548447

En este caso podemo ver que la máxima diferencia entre el valor real y el del modelo es de 5208.81 dolares

Bibliografia Y Referencias:

* Todos los datos de los precios del Bitcoiny las commodities provienen de la página https://www.investing.com
* Se usó la siguiente documentación: 
 * https://plotly.com/julia/figure-labels/
 * http://juliaplots.org/PlotlyJS.jl/stable/examples/line_scatter/
 * https://discourse.julialang.org/t/plotlyjs-how-to-change-the-default-labels-of-variables/60798